### PR TITLE
Allow decoding in sub-batches

### PR DESCRIPTION
**Pull request by kijai**
This allows decoding in sub-batches for processing large batches such as videos:

image

### CODE REVIEW

**Review by huchenlei — COMMENTED** [review]
NVM, it is a problem with `Apply Mask to Image` node before the PR.

I get following error when running the example workflow on `Apply Mask to Image` node:

```
Error occurred when executing ETN_ApplyMaskToImage:

expand(torch.HalfTensor{[32, 1024, 1024]}, size=[1024, 1024]): the number of sizes provided (2) must be greater or equal to the number of dimensions in the tensor (3)

File "D:\ComfyUI_windows_portable\ComfyUI\execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "D:\ComfyUI_windows_portable\ComfyUI\execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "D:\ComfyUI_windows_portable\ComfyUI\execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "D:\ComfyUI_windows_portable\ComfyUI\custom_nodes\comfyui-tooling-nodes\nodes.py", line 137, in apply_mask
out[i, 3, :, :] = mask
```

File: layered_diffusion.py [hunk 1]
```diff
@@ -68,14 +74,17 @@ def decode(self, samples, images):
...
+        for start_idx in range(0, samples["samples"].shape[0], per_batch):
+            decoded.append(self.vae_transparent_decoder.decode_pixel(pixel[start_idx:start_idx+per_batch], samples["samples"][start_idx:start_idx+per_batch]))
+        pixel_with_alpha = torch.cat(decoded, dim=0)
+        
         # [B, C, H, W] => [B, H, W, C]
         pixel_with_alpha = pixel_with_alpha.movedim(1, -1)
         image = pixel_with_alpha[..., 1:]
         alpha = pixel_with_alpha[..., 0]
-        return (image, alpha)
+        return (image.cpu(), alpha.cpu(),)
```
Comment (huchenlei) [inline layered_diffusion.py:87]: Why do we need to move tensors to cpu here?

**Review by huchenlei — CHANGES_REQUESTED** [review]
If am understanding this correctly, we are trying to reduce vram consumption by dividing tasks into sub-batches?

File: layered_diffusion.py (hunk 1 above)
Comment (kijai) [reply layered_diffusion.py]: That is not necessary indeed, there was some node that expected it on cpu when I was testing, removed it now.

**Comment by kijai** [comment]
> If am understanding this correctly, we are trying to reduce vram consumption by dividing tasks into sub-batches?

Yeah, we have such nodes for normal encoding/decoding already, the VRAM spike otherwise can be huge with large batches common with videos.

**Comment by huchenlei** [comment]
I am still undecided if this is really necessary to be added to the decode node. If other parts of the pipeline are handling the full batch, why the decode unet need to make compromise and perform decode in smaller batch.

I wonder if there are existing nodes to split large video (image seq) down to small batches?

**Comment by kijai** [comment]
> I am still undecided if this is really necessary to be added to the decode node. If other parts of the pipeline are handling the full batch, why the decode unet need to make compromise and perform decode in smaller batch.
> 
> I wonder if there are existing nodes to split large video (image seq) down to small batches?

Because it's not possible to decode a large batch at once. It will OOM, and we can't use other decoders as this node is only one with the compatible decoder. With AnimateDiff we have to use complete batches and it can't be split up, as it uses sliding context windows to handle large batches. 

It could just as well be a separate too in this repo. As far as I know I can't import the VAE used here to my custom node for example, and as this node loads the VAE directly I'd have to add all that as well, hence the PR.

File: layered_diffusion.py (hunk 1 above)
Comment (huchenlei) [reply layered_diffusion.py]: They do seem necessary though. See https://github.com/huchenlei/ComfyUI-layerdiffuse/issues/28.

Comment (kijai) [reply layered_diffusion.py]: Honestly there doesn't seem to be a consensus which way it's supposed to be in Comfy, I do think .cpu() can be safer as most stuff that need it in gpu convert it first anyway.

**Review by huchenlei — APPROVED** [review]
